File: examples/tailcat-e2b/python/tailcat_e2b/tailcat_runtime.py
"""Helpers for running Tailcat inside E2B sandboxes and on this machine.

A Tailcat *server* (listener) prints an address; a Tailcat *client* connects to
it. Both sides can live in a sandbox or on the laptop. This module hides the
three things that are easy to get wrong: reading the address, waiting until the
server is actually reachable, and turning on direct paths.
"""

from __future__ import annotations

import hashlib
import os
import re
import shutil
import subprocess
import sys
import tempfile
import time
from dataclasses import dataclass
from typing import Callable

from e2b import CommandExitException, Sandbox

from .template import ENABLE_DIRECT_PATHS_CMD, TAILCAT_URL, TEMPLATE_ALIAS

# A Tailcat address looks like `tco2FwWCD...`: the prefix `tc` followed by base64url.
TAILCAT_ADDRESS_PATTERN = re.compile(r"\btc[A-Za-z0-9_-]{20,}")

INSTALL_TAILCAT_CMD = (
    f"curl -fsSL -o /tmp/tailcat.tgz {TAILCAT_URL} && mkdir -p /tmp/tc && tar xzf /tmp/tailcat.tgz -C /tmp/tc"
    " && sudo install -m 0755 $(find /tmp/tc -type f -name tailcat) /usr/local/bin/tailcat && rm -rf /tmp/tc /tmp/tailcat.tgz"
)

# Keep our prints in order with the output of the tailcat subprocesses.
sys.stdout.reconfigure(line_buffering=True)


@dataclass
class CommandResult:
    exit_code: int
    output: str
    """stdout and stderr, concatenated."""


CommandRunner = Callable[[str], CommandResult]
"""Runs a shell command locally or in a sandbox and captures its exit code and combined output."""


@dataclass
class TailcatServer:
    """A running Tailcat server and how to stop it."""

    address: str
    stop: Callable[[], None]


# ---------------------------------------------------------------------------
# Sandboxes
# ---------------------------------------------------------------------------


def create_sandbox(timeout_seconds: int = 600) -> Sandbox:
    """Create a sandbox with Tailcat installed and direct paths enabled."""
    started_at = time.time()
    sandbox = _create_sandbox_from_template_or_fallback(timeout_seconds)
    print(f"[{sandbox.sandbox_id}] sandbox ready in {time.time() - started_at:.1f}s")

    # Idempotent: the template's start command already does this, but sandboxes
    # made from other templates need it too. See README, "Quirks".
    sandbox.commands.run(ENABLE_DIRECT_PATHS_CMD)
    return sandbox


def _create_sandbox_from_template_or_fallback(timeout_seconds: int) -> Sandbox:
    """Use the `tailcat` template when this team has built it, else a default sandbox with Tailcat installed on the fly."""
    try:
        return Sandbox.create(TEMPLATE_ALIAS, timeout=timeout_seconds)
    except Exception as error:
        template_missing = "404" in str(error)
        if not template_missing:
            raise
    sandbox = Sandbox.create(timeout=timeout_seconds)
    sandbox.commands.run(INSTALL_TAILCAT_CMD, timeout=120)
    print(
        f"[{sandbox.sandbox_id}] template '{TEMPLATE_ALIAS}' not found, installed tailcat at runtime instead "
        "(run `poetry run build-template` once to skip this)"
    )
    return sandbox


def run_sandbox_command(sandbox: Sandbox, command: str, timeout_seconds: int = 60) -> CommandResult:
    """Run a command in the sandbox without raising on a non-zero exit."""
    try:
        result = sandbox.commands.run(command, timeout=timeout_seconds)
        return CommandResult(result.exit_code, result.stdout + result.stderr)
    except CommandExitException as error:
        return CommandResult(error.exit_code, error.stdout + error.stderr)


def create_sandbox_command_runner(sandbox: Sandbox, timeout_seconds: int = 120) -> CommandRunner:
    """Create a command runner bound to one sandbox."""

    def run_in_sandbox(command: str) -> CommandResult:
        return run_sandbox_command(sandbox, command, timeout_seconds)

    return run_in_sandbox


def local_command_runner(command: str) -> CommandResult:
    completed_process = subprocess.run(command, shell=True, capture_output=True, text=True)
    return CommandResult(completed_process.returncode, completed_process.stdout + completed_process.stderr)


def require_successful_command(result: CommandResult) -> None:
    """Raise with the command output when a command did not exit 0."""
    if result.exit_code != 0:
        raise RuntimeError(f"command failed (exit {result.exit_code}):\n{result.output}")


# ---------------------------------------------------------------------------
# Tailcat servers
# ---------------------------------------------------------------------------


def start_sandbox_tailcat_server(
    sandbox: Sandbox,
    tailcat_arguments: str,
    instance_name: str = "tailcat",
    wait_seconds: float = 20,
    stdout_file: str | None = None,
) -> TailcatServer:
    """Start `tailcat <tailcat_arguments>` in the background inside the sandbox and return its address.

    `tailcat_arguments` is passed to the shell unsplit. Tailcat writes the
    address to TAILCAT_ADDR_FILE, which is more reliable than scraping stderr
    from a background command. `stdout_file` redirects the server's stdout
    (bare `tailcat` copies its one connection there).
    """
    address_file = f"/tmp/{instance_name}.addr"
    log_file = f"/tmp/{instance_name}.log"
    server_process = sandbox.commands.run(
        f"rm -f {address_file}; TAILCAT_ADDR_FILE={address_file} tailcat {tailcat_arguments} > {stdout_file or log_file} 2> {log_file}",
        background=True,
    )

    def read_address_file() -> str:
        return run_sandbox_command(sandbox, f"cat {address_file} 2>/dev/null").output

    def read_server_log() -> str:
        return run_sandbox_command(sandbox, f"cat {log_file}").output

    def stop() -> None:
        server_process.kill()

    address = _wait_for_tailcat_address(
        read_address_file=read_address_file,
        wait_seconds=wait_seconds,
        read_server_log=read_server_log,
    )
    print(f"[{sandbox.sandbox_id}] tailcat {tailcat_arguments}  ->  {_abbreviate(address)}")
    return TailcatServer(address, stop=stop)


def start_local_tailcat_server(tailcat_arguments: str, wait_seconds: float = 20) -> TailcatServer:
    """Start `tailcat <tailcat_arguments>` on this machine and return its address. `tailcat_arguments` is split on whitespace."""
    require_local_tailcat()
    address_file = os.path.join(tempfile.gettempdir(), f"tailcat-{os.getpid()}-{int(time.time() * 1000)}.addr")
    server_process = subprocess.Popen(
        ["tailcat", *tailcat_arguments.split()],
        env={**os.environ, "TAILCAT_ADDR_FILE": address_file},
        stdout=subprocess.DEVNULL,
        stderr=subprocess.DEVNULL,
    )

    def read_address_file() -> str:
        if not os.path.exists(address_file):
            return ""
        with open(address_file) as file:
            return file.read()

    def stop() -> None:
        server_process.terminate()
        if os.path.exists(address_file):
            os.unlink(address_file)

    try:
        address = _wait_for_tailcat_address(read_address_file, wait_seconds)
    except Exception:
        server_process.terminate()
        raise
    print(f"[laptop] tailcat {tailcat_arguments}  ->  {_abbreviate(address)}")
    return TailcatServer(address, stop)


def _wait_for_tailcat_address(
    read_address_file: Callable[[], str],
    wait_seconds: float,
    read_server_log: Callable[[], str] | None = None,
) -> str:
    """Poll `read_address_file` until it contains a Tailcat address."""
    started_at = time.time()
    while time.time() - started_at < wait_seconds:
        if address_match := TAILCAT_ADDRESS_PATTERN.search(read_address_file()):
            return address_match.group(0)
        time.sleep(0.3)
    server_log = f"\n{read_server_log()}" if read_server_log else ""
    raise RuntimeError(f"tailcat did not print an address within {wait_seconds}s{server_log}")


# ---------------------------------------------------------------------------
# Tailcat clients
# ---------------------------------------------------------------------------


def wait_until_reachable(run_command: CommandRunner, address: str, timeout_seconds: float = 30) -> None:
    """Block until a Tailcat server answers a ping from the client side.

    A Tailcat server prints its address before it has finished connecting to
    the DERP relay. A client that connects in that window is told the relay
    does not know the peer, so retry instead of trusting the first attempt.
    """
    started_at = time.time()
    while time.time() - started_at < timeout_seconds:
        ping = run_command(f"tailcat ping --timeout 3s {address}")
        if ping.exit_code == 0:
            return
        time.sleep(1)
    raise RuntimeError(f"tailcat server {_abbreviate(address)} not reachable after {timeout_seconds}s")


def describe_network_path(run_command: CommandRunner, address: str, timeout: str = "20s") -> str:
    """Ping until a direct path forms and describe the result: relay latency, direct latency, or neither."""
    ping = run_command(f"tailcat ping --until-direct --timeout {timeout} {address}")
    pong_lines = [line for line in ping.output.splitlines() if "pong in" in line]
    relay_latency = next((line.split("pong in ")[1] for line in pong_lines if "DERP" in line), None)
    direct_latency = next((line.split("pong in ")[1] for line in pong_lines if "DERP" not in line), None)
    if direct_latency:
        return f"direct path: {direct_latency}" + (f" (relay was {relay_latency})" if relay_latency else "")
    if relay_latency:
        return f"relay only: {relay_latency} (no direct path within {timeout})"
    return f"no pong: {ping.output.strip()[-200:]}"


# ---------------------------------------------------------------------------
# Small utilities
# ---------------------------------------------------------------------------


def require_local_tailcat() -> None:
    if not shutil.which("tailcat"):
        raise SystemExit(
            "tailcat is not installed on this machine. Install it with `brew install tailcat` "
            "or see https://github.com/tailscale/tailcat#install"
        )


def compute_local_md5(path: str) -> str:
    digest = hashlib.md5()
    with open(path, "rb") as file:
        while chunk := file.read(1 << 20):
            digest.update(chunk)
    return digest.hexdigest()


def compute_sandbox_md5(sandbox: Sandbox, path: str) -> str:
    result = run_sandbox_command(sandbox, f"md5sum {path}")
    require_successful_command(result)
    return result.output.split()[0]


def format_megabits_per_second(size_bytes: int, seconds: float) -> str:
    return f"{size_bytes * 8 / seconds / 1e6:.0f} Mbit/s"


def seconds_since(started_at: float) -> float:
    return time.time() - started_at


def _abbreviate(address: str) -> str:
    return f"{address[:20]}…"

File: examples/tailcat-e2b/js/src/tailcatRuntime.ts
/**
 * Helpers for running Tailcat inside E2B sandboxes and on this machine.
 *
 * A Tailcat *server* (listener) prints an address; a Tailcat *client* connects
 * to it. Both sides can live in a sandbox or on the laptop. This module hides
 * the two things that are easy to get wrong: reading the address and waiting
 * until the server is actually reachable through the DERP relay.
 */
import { spawn, spawnSync, type ChildProcess } from "node:child_process";
import { createHash } from "node:crypto";
import { existsSync, readFileSync, unlinkSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { setTimeout as delay } from "node:timers/promises";
import { CommandExitError, Sandbox } from "e2b";
import { TAILCAT_URL, TEMPLATE_ALIAS } from "../template/template";

/** A Tailcat address looks like `tco2FwWCD...`: the prefix `tc` followed by base64url. */
const TAILCAT_ADDRESS_PATTERN = /\btc[A-Za-z0-9_-]{20,}/;

const INSTALL_TAILCAT_CMD =
  `curl -fsSL -o /tmp/tailcat.tgz ${TAILCAT_URL} && mkdir -p /tmp/tc && tar xzf /tmp/tailcat.tgz -C /tmp/tc` +
  " && sudo install -m 0755 $(find /tmp/tc -type f -name tailcat) /usr/local/bin/tailcat && rm -rf /tmp/tc /tmp/tailcat.tgz";

export interface CommandResult {
  exitCode: number;
  /** stdout and stderr, concatenated. */
  output: string;
}

/** Runs a shell command locally or in a sandbox and captures its exit code and combined output. */
export type CommandRunner = (command: string) => Promise<CommandResult>;

/** A running Tailcat server and how to stop it. */
export interface TailcatServer {
  address: string;
  stop(): Promise<void> | void;
}

// ---------------------------------------------------------------------------
// Sandboxes
// ---------------------------------------------------------------------------

/** Create a sandbox with Tailcat installed. */
export async function createSandbox(timeoutMs = 600_000): Promise<Sandbox> {
  const startedAt = Date.now();
  const sandbox = await createSandboxFromTemplateOrFallback(timeoutMs);
  const elapsedSeconds = (Date.now() - startedAt) / 1000;
  console.log(
    `[${sandbox.sandboxId}] sandbox ready in ${elapsedSeconds.toFixed(1)}s`,
  );

  return sandbox;
}

/** Use the `tailcat` template when this team has built it, else a default sandbox with Tailcat installed on the fly. */
async function createSandboxFromTemplateOrFallback(
  timeoutMs: number,
): Promise<Sandbox> {
  try {
    return await Sandbox.create(TEMPLATE_ALIAS, { timeoutMs });
  } catch (error) {
    const templateMissing = String(error).includes("404");
    if (!templateMissing) {
      throw error;
    }
  }
  const sandbox = await Sandbox.create({ timeoutMs });
  await sandbox.commands.run(INSTALL_TAILCAT_CMD, { timeoutMs: 120_000 });
  console.log(
    `[${sandbox.sandboxId}] template '${TEMPLATE_ALIAS}' not found, installed tailcat at runtime instead ` +
      "(run `npm run e2b:build` once to skip this)",
  );
  return sandbox;
}

/** Run a command in the sandbox without throwing on a non-zero exit. */
export async function runSandboxCommand(
  sandbox: Sandbox,
  command: string,
  timeoutMs = 60_000,
): Promise<CommandResult> {
  try {
    const result = await sandbox.commands.run(command, { timeoutMs });
    return { exitCode: result.exitCode, output: result.stdout + result.stderr };
  } catch (error) {
    if (error instanceof CommandExitError) {
      return { exitCode: error.exitCode, output: error.stdout + error.stderr };
    }
    throw error;
  }
}

/** Create a command runner bound to one sandbox. */
export function createSandboxCommandRunner(
  sandbox: Sandbox,
  timeoutMs = 120_000,
): CommandRunner {
  return async function runInSandbox(command: string): Promise<CommandResult> {
    return runSandboxCommand(sandbox, command, timeoutMs);
  };
}

export function localCommandRunner(command: string): Promise<CommandResult> {
  const completedProcess = spawnSync(command, {
    shell: true,
    encoding: "utf8",
  });
  return Promise.resolve({
    exitCode: completedProcess.status ?? 1,
    output: (completedProcess.stdout ?? "") + (completedProcess.stderr ?? ""),
  });
}

/** Throw with the command output when a command did not exit 0. */
export function requireSuccessfulCommand(result: CommandResult): void {
  if (result.exitCode !== 0) {
    throw new Error(
      `command failed (exit ${result.exitCode}):\n${result.output}`,
    );
  }
}

// ---------------------------------------------------------------------------
// Tailcat servers
// ---------------------------------------------------------------------------

/**
 * Start `tailcat <tailcatArguments>` in the background inside the sandbox and return its address.
 *
 * `tailcatArguments` is passed to the shell unsplit. Tailcat writes the address
 * to TAILCAT_ADDR_FILE, which is more reliable than scraping stderr from a
 * background command. `stdoutFile` redirects the server's stdout (bare
 * `tailcat` copies its one connection there).
 */
export async function startSandboxTailcatServer(
  sandbox: Sandbox,
  tailcatArguments: string,
  instanceName = "tailcat",
  {
    waitMs = 20_000,
    stdoutFile,
  }: { waitMs?: number; stdoutFile?: string } = {},
): Promise<TailcatServer> {
  const addressFile = `/tmp/${instanceName}.addr`;
  const logFile = `/tmp/${instanceName}.log`;
  const serverProcess = await sandbox.commands.run(
    `rm -f ${addressFile}; TAILCAT_ADDR_FILE=${addressFile} tailcat ${tailcatArguments} > ${
      stdoutFile ?? logFile
    } 2> ${logFile}`,
    { background: true },
  );

  async function readAddressFile(): Promise<string> {
    return (await runSandboxCommand(sandbox, `cat ${addressFile} 2>/dev/null`))
      .output;
  }

  async function readServerLog(): Promise<string> {
    return (await runSandboxCommand(sandbox, `cat ${logFile}`)).output;
  }

  async function stop(): Promise<void> {
    await serverProcess.kill();
  }

  const address = await waitForTailcatAddress(
    readAddressFile,
    waitMs,
    readServerLog,
  );
  console.log(
    `[${sandbox.sandboxId}] tailcat ${tailcatArguments}  ->  ${abbreviate(
      address,
    )}`,
  );
  return { address, stop };
}

/** Start `tailcat <tailcatArguments>` on this machine and return its address. `tailcatArguments` is split on whitespace. */
export async function startLocalTailcatServer(
  tailcatArguments: string,
  waitMs = 20_000,
): Promise<TailcatServer> {
  requireLocalTailcat();
  const addressFile = join(
    tmpdir(),
    `tailcat-${process.pid}-${Date.now()}.addr`,
  );
  const serverProcess: ChildProcess = spawn(
    "tailcat",
    tailcatArguments.split(/\s+/).filter(Boolean),
    {
      env: { ...process.env, TAILCAT_ADDR_FILE: addressFile },
      stdio: "ignore",
    },
  );

  function readAddressFile(): Promise<string> {
    const contents = existsSync(addressFile)
      ? readFileSync(addressFile, "utf8")
      : "";
    return Promise.resolve(contents);
  }

  let address: string;
  try {
    address = await waitForTailcatAddress(readAddressFile, waitMs);
  } catch (error) {
    serverProcess.kill();
    throw error;
  }
  console.log(
    `[laptop] tailcat ${tailcatArguments}  ->  ${abbreviate(address)}`,
  );

  function stop(): void {
    serverProcess.kill();
    if (existsSync(addressFile)) {
      unlinkSync(addressFile);
    }
  }

  return {
    address,
    stop,
  };
}

/** Poll `readAddressFile` until it contains a Tailcat address. */
async function waitForTailcatAddress(
  readAddressFile: () => Promise<string>,
  waitMs: number,
  readServerLog?: () => Promise<string>,
): Promise<string> {
  const startedAt = Date.now();
  while (Date.now() - startedAt < waitMs) {
    const addressMatch = TAILCAT_ADDRESS_PATTERN.exec(await readAddressFile());
    if (addressMatch) {
      return addressMatch[0];
    }
    await delay(300);
  }
  const serverLog = readServerLog ? `\n${await readServerLog()}` : "";
  throw new Error(
    `tailcat did not print an address within ${waitMs / 1000}s${serverLog}`,
  );
}

// ---------------------------------------------------------------------------
// Tailcat clients
// ---------------------------------------------------------------------------

/**
 * Block until a Tailcat server answers a ping from the client side.
 *
 * A Tailcat server prints its address before it has finished connecting to the
 * DERP relay. A client that connects in that window is told the relay does not
 * know the peer, so retry instead of trusting the first attempt.
 */
export async function waitUntilReachable(
  runCommand: CommandRunner,
  address: string,
  timeoutMs = 30_000,
): Promise<void> {
  const startedAt = Date.now();
  while (Date.now() - startedAt < timeoutMs) {
    const ping = await runCommand(`tailcat ping --timeout 3s ${address}`);
    if (ping.exitCode === 0) {
      return;
    }
    await delay(1000);
  }
  throw new Error(
    `tailcat server ${abbreviate(address)} not reachable after ${
      timeoutMs / 1000
    }s`,
  );
}

/** Ping once and describe the Tailcat connection. */
export async function describeConnection(
  runCommand: CommandRunner,
  address: string,
  timeout = "5s",
): Promise<string> {
  const ping = await runCommand(`tailcat ping --timeout ${timeout} ${address}`);
  const pongLine = ping.output
    .split("\n")
    .find((line) => line.includes("pong in"));
  if (!pongLine) {
    return `no pong: ${ping.output.trim().slice(-200)}`;
  }
  const latency = pongLine.split("pong in ")[1] ?? pongLine.trim();
  return pongLine.includes("DERP")
    ? `DERP relay: ${latency}`
    : `Tailcat connection: ${latency}`;
}

// ---------------------------------------------------------------------------
// Small utilities
// ---------------------------------------------------------------------------

export function requireLocalTailcat(): void {
  if (spawnSync("tailcat", ["version"], { stdio: "ignore" }).status !== 0) {
    console.error(
      "tailcat is not installed on this machine. Install it with `brew install tailcat` or see https://github.com/tailscale/tailcat#install",
    );
    process.exit(1);
  }
}

export function computeLocalMd5(path: string): string {
  return createHash("md5").update(readFileSync(path)).digest("hex");
}

export async function computeSandboxMd5(
  sandbox: Sandbox,
  path: string,
): Promise<string> {
  const result = await runSandboxCommand(sandbox, `md5sum ${path}`);
  requireSuccessfulCommand(result);
  return result.output.trim().split(/\s+/)[0] ?? "";
}

export function formatMegabitsPerSecond(
  bytes: number,
  seconds: number,
): string {
  return `${Math.round((bytes * 8) / seconds / 1e6)} Mbit/s`;
}

export function secondsSince(startedAt: number): number {
  return (Date.now() - startedAt) / 1000;
}

function abbreviate(address: string): string {
  return `${address.slice(0, 20)}…`;
}
